ocaml/src/alba_osd.ml

```diff
@@ -0,0 +1,281 @@
+                    then Lwt.return (Nsm_model.Assert.ObjectHasId
+                                       (object_name,
+                                        mf.Nsm_model.Manifest.object_id))
+                    else Lwt.fail Nsm_model.Err.(Nsm_exn (Assert_failed, object_name)))
+            asserts
+        in
+
+        Lwt.catch
+          (fun () ->
+           let nsm_asserts_t = prepare_asserts () in
+           let nsm_updates_t =
+             alba_client # nsm_host_access # with_namespace_id
+                         ~namespace
+                         prepare_updates
+           in
+           nsm_asserts_t >>= fun nsm_asserts ->
+           nsm_updates_t >>= fun nsm_updates ->
+
+           alba_client # nsm_host_access
+                       # with_nsm_client
+                       ~namespace
+                       (fun nsm ->
+                        nsm # apply_sequence nsm_asserts nsm_updates) >>= fun () ->
+
+           (* update the manifest cache *)
+           alba_client # nsm_host_access # with_namespace_id
+                       ~namespace
+                       (fun namespace_id ->
+                        List.iter
+                          (function
+                            | Nsm_model.Update.DeleteObject _ -> ()
+                            | Nsm_model.Update.PutObject (mf, _) ->
+                               Manifest_cache.ManifestCache.add
+                                 (alba_client # get_manifest_cache)
+                                 namespace_id mf.Nsm_model.Manifest.name mf)
+                          nsm_updates;
+                        Lwt.return ())
+           >>= fun () ->
+           Lwt.return Osd.Ok)
+          (function
+            | Nsm_model.Err.Nsm_exn (Nsm_model.Err.Assert_failed, key) ->
+               Lwt.return (Osd.Exn Osd.Error.(Assert_failed key))
+            | exn ->
+               Lwt.fail exn)
+    end
+  in
+  object(self :# Osd.osd)
+
+    method global_kvs =
+      get_kvs ~consistent_read:true prefix
+
+    method namespace_kvs namespace_id =
+      get_kvs ~consistent_read:false (to_namespace_name namespace_id)
+
+    method add_namespace namespace_id =
+      alba_client # mgr_access # get_namespace
+                  ~namespace:(to_namespace_name namespace_id) >>= fun r ->
+      if (r <> None)
+      then Lwt.return ()
+      else alba_client # create_namespace
+                       ~namespace:(to_namespace_name namespace_id)
+                       ~preset_name
+                       () >>= fun _ ->
+           Lwt.return ()
+
+    method delete_namespace namespace_id _ =
+      alba_client # delete_namespace
+                  ~namespace:(to_namespace_name namespace_id) >>= fun () ->
+      Lwt.return None
+
+    method set_full _ = failwith "grmbl this method doesn't belong here."
+    method get_version = Lwt.return Alba_version.summary
+    method get_long_id = alba_id
+    method get_disk_usage = Lwt.return
```

Comment:
the name should (in time) change. the implementation for alternatives can be done in a later pr. This one's large enough already.